### PR TITLE
Interpreter: replay and reload

### PR DESCRIPTION
**Pull request by asterite**
This is one way to workaround #12144 which I think works quite well.

The idea is this: when we detect structural changes (a class is defined or reopened, a method is defined, etc.) then we replay the entire session by incorporating that change.

The only not-so-nice thing about this is that if the past inputs had side-effects, for example writing to a file, asking the current time, writing a DB record, etc., then those would be performed again. So in a way this makes it work like the old `icr`, but only when there are structural changes. But I think it's not that bad, and it might actually make sense!

For example... imagine that you have some code that writes a file in some way. But some of the functions you used are buggy, or you'd like to change them. So you do so, the history is replayed, and you can see the new output right away.

(or maybe the last paragraph is me trying to find an excuse for this behavior)

But there are more reasons... consider this session and see what happens:

crystal-i-replay

We reopened `Point` and added a new instance variable with a default value. That's perfectly fine because it will be automatically initialized in all constructors. But what would be the value for the existing `point` instance we declared in the past? No problem! Because the program is replayed, that value is computed there.

So that's why I think it kind of makes sense to do it like this.

What happens if the structural change is impossible to apply? Consider this now:

crystal-i-replay-2

We added an instance variable to `Point` without a default value... it must be initialized in all the `initialize`, but we didn't do that... plus we already created a point in the past! So here we just discard the recent change. It's similar to discarding the last input if it had a syntax error, or a semantic error. I believe that if this happens it's probably a mistake, you didn't intend to add a non-nilable field that can't be initialized.

The nice thing about replaying stuff is that it will actually reparse and reload files that you `require`d. So you could `require` a file, write some repl code, then change the required files and accidentally introduce a structural change so it's reloaded!

But that's tedious to do, so I also added a `reload!` command, similar to `pry` (only works in the repl, not in `pry`!). Check this out:

crystal-i-replay-3

In this way you can write complex logic in your favorite editor and have those changes be there in the repl. Or you can write them in the repl if you want.

### CODE REVIEW

**Comment by asterite** [comment]
Note: I didn't try it with `require "http/client"` because that still doesn't work for me on mac ("that licrypto loaded in an unsafe way" thing) but I'm almost sure it should work. Let me know if someone gets a chance to try it.

File: src/compiler/crystal/interpreter/repl.cr [hunk 1]
```diff
@@ -94,22 +115,82 @@ class Crystal::Repl
...
+      # don't consider it for future replays.
+      @past_nodes.pop
+      raise ex
+    end
+
     @interpreter.interpret(node, @main_visitor.meta_vars)
   end
 
+  private def handle_structural_changes(show_output : Bool = true)
+    puts "Structural change detected. Replyaing session...", &.dark_gray if show_output
```
Comment (Sija) [inline src/compiler/crystal/interpreter/repl.cr:145]:
```suggestion
    puts "Structural change detected. Replaying session...", &.dark_gray if show_output
```

File: src/compiler/crystal/interpreter/repl.cr [hunk 2]
```diff
@@ -94,22 +115,82 @@ class Crystal::Repl
...
+
+    begin
+      value = replay
+      puts "Session replayed!", &.dark_gray if show_output
+      value
+    rescue ex : Crystal::Error
+      puts "An error happened while replyaing with the recent structural changes:", &.yellow
```
Comment (Sija) [inline src/compiler/crystal/interpreter/repl.cr:152]:
```suggestion
      puts "An error happened while replaying with the recent structural changes:", &.yellow
```

Comment (asterite) [reply src/compiler/crystal/interpreter/repl.cr:152]: Wow! It's not an accident then, it's encoded in my brain 😮

File: src/compiler/crystal/codegen/link.cr [hunk 3]
```diff
@@ -209,7 +209,11 @@ module Crystal
       types.try &.each_value do |type|
         next if type.is_a?(AliasType) || type.is_a?(TypeDefType)
 
-        if type.is_a?(LibType) && type.used? && (link_annotations = type.link_annotations)
+        if type.is_a?(LibType) &&
+           # In interpreted mode we want to load a library even if it's not used,
+           # because it might be used later on, interactively.
+           (interpreted? || type.used?) &&
```
Comment (straight-shoota) [inline src/compiler/crystal/codegen/link.cr:215]: I don't think this is a good idea. Interpreted programs should behave as much as compiled ones as possible.
When loading all libraries unconditionally the loader will error on any missing lib even if it's not even used. In a compiled program libraries are only linked when they're actually used. So you might get a program that runs fine in compiled mode, but fails in interpreted mode (for no reason).

Comment (asterite) [reply src/compiler/crystal/codegen/link.cr:215]: Okay, I'll undo this change.

**Comment by asterite** [comment]
Maybe this isn't the best way to do this.

**Comment by straight-shoota** [comment]
@asterite Could you explain your reasoning maybe? (don't need to do this right away, just for the rest of us to follow it)

**Pull request closed without merging**